migrations/versions/0465_delete_broadcast_data.py

```diff
@@ -0,0 +1,104 @@
+            "DELETE FROM api_keys WHERE service_id in :service_ids;",
+            "DELETE FROM complaints WHERE service_id in :service_ids;",
+            "DELETE FROM service_sms_senders WHERE service_id in :service_ids;",
+            "DELETE FROM inbound_numbers WHERE service_id in :service_ids;",
+            "DELETE FROM inbound_sms WHERE service_id in :service_ids;",
+            "DELETE FROM inbound_sms_history WHERE service_id in :service_ids;",
+            "DELETE FROM invited_users WHERE service_id in :service_ids;",
+            "DELETE FROM jobs WHERE service_id in :service_ids;",
+            "DELETE FROM notification_history WHERE service_id in :service_ids;",
+            "DELETE FROM notifications WHERE service_id in :service_ids;",
+            "DELETE FROM permissions WHERE service_id in :service_ids;",
+            "DELETE FROM returned_letters WHERE service_id in :service_ids;",
+            "DELETE FROM service_callback_api WHERE service_id in :service_ids;",
+            "DELETE FROM service_contact_list WHERE service_id in :service_ids;",
+            "DELETE FROM service_data_retention WHERE service_id in :service_ids;",
+            "DELETE FROM service_email_branding WHERE service_id in :service_ids;",
+            "DELETE FROM service_email_reply_to WHERE service_id in :service_ids;",
+            "DELETE FROM service_inbound_api WHERE service_id in :service_ids;",
+            "DELETE FROM service_join_requests WHERE service_id in :service_ids;",
+            "DELETE FROM service_letter_branding WHERE service_id in :service_ids;",
+            "DELETE FROM service_permissions WHERE service_id in :service_ids;",
+            "DELETE FROM service_whitelist WHERE service_id in :service_ids;",
+            "DELETE FROM user_folder_permissions WHERE service_id in :service_ids",
+            "DELETE FROM template_folder_map WHERE template_id in (SELECT id FROM templates WHERE service_id in :service_ids);",
+            "DELETE FROM template_redacted WHERE template_id in (SELECT id FROM templates WHERE service_id in :service_ids);",
+            "DELETE FROM template_folder WHERE service_id in :service_ids;",
+            "DELETE FROM templates WHERE service_id in :service_ids;",
+            "DELETE FROM templates_history WHERE service_id in :service_ids;",
+            "DELETE FROM service_letter_contacts WHERE service_id in :service_ids;",
+            "DELETE FROM unsubscribe_request WHERE service_id in :service_ids;",
+            "DELETE FROM unsubscribe_request_history WHERE service_id in :service_ids;",
+            "DELETE FROM unsubscribe_request_report WHERE service_id in :service_ids;",
+            "DELETE FROM user_to_service WHERE service_id in :service_ids;",
+            # now we can finally delete the services
+            "DELETE FROM services WHERE id in :service_ids;",
+            "DELETE FROM services_history WHERE id in :service_ids;",
+        ]
+
+    for delete_statement in delete_statements:
```

Comment:
Good question!

Alembic runs everything in a transaction - this is pretty important so that if a line fails, it can roll back (transactions will also roll back schema changes like changing tables etc too), so that the `alembic_version` is consistent with the actual state of the table. Otherwise you'd have to unpick everything manually which would be rubbish.

You can see where we pass this flag to alembic here: https://github.com/alphagov/notifications-api/blob/main/migrations/env.py#L91

If you run a migration locally you'll see the logs reflecting this too - eg see the logs below which start with "BEGIN (implicit)" and end with COMMIT

<details><summary>Example log</summary>
<p>
```
INFO  [sqlalchemy.engine.Engine] BEGIN (implicit)
INFO  [alembic.runtime.migration] Running upgrade 0463_backfill_created_at -> 0464_create_svc_join_requests
INFO  [sqlalchemy.engine.Engine] CREATE TYPE request_status AS ENUM ('pending', 'approved', 'rejected')
INFO  [sqlalchemy.engine.Engine] [no key 0.00005s] {}
INFO  [sqlalchemy.engine.Engine]
CREATE TABLE service_join_requests (
	id UUID NOT NULL,
	requester_id UUID NOT NULL,
	service_id UUID NOT NULL,
	created_at TIMESTAMP WITHOUT TIME ZONE DEFAULT now() NOT NULL,
	status request_status DEFAULT 'pending' NOT NULL,
	status_changed_at TIMESTAMP WITHOUT TIME ZONE,
	status_changed_by_id UUID,
	reason TEXT,
	PRIMARY KEY (id),
	FOREIGN KEY(requester_id) REFERENCES users (id),
	FOREIGN KEY(service_id) REFERENCES services (id),
	FOREIGN KEY(status_changed_by_id) REFERENCES users (id)
)
INFO  [sqlalchemy.engine.Engine] [no key 0.00008s] {}
INFO  [sqlalchemy.engine.Engine]
CREATE TABLE contacted_users (
	service_join_request_id UUID NOT NULL,
	user_id UUID NOT NULL,
	PRIMARY KEY (service_join_request_id, user_id),
	FOREIGN KEY(service_join_request_id) REFERENCES service_join_requests (id),
	FOREIGN KEY(user_id) REFERENCES users (id)
)
INFO  [sqlalchemy.engine.Engine] [no key 0.00008s] {}
INFO  [sqlalchemy.engine.Engine] UPDATE alembic_version SET version_num='0464_create_svc_join_requests' WHERE alembic_version.version_num = '0463_backfill_created_at'
INFO  [sqlalchemy.engine.Engine] [generated in 0.00008s] {}
INFO  [sqlalchemy.engine.Engine] COMMIT
```

</p>
</details>

Reply:
nice, that's very useful. thanks, I did not know!